pkg/sql/opt/optbuilder/scope.go

```diff
@@ -476,9 +476,11 @@ func (s *scope) resolveType(expr tree.Expr, desired *types.T) tree.TypedExpr {
 // expression with no error). If the result type is types.Unknown, then
 // resolveType will wrap the expression in a type cast in order to produce the
 // desired type.
-func (s *scope) resolveAndRequireType(expr tree.Expr, desired *types.T) tree.TypedExpr {
+func (s *scope) resolveAndRequireType(
+	expr tree.Expr, desired *types.T, mutSuffix string,
+) tree.TypedExpr {
 	expr = s.walkExprTree(expr)
-	texpr, err := tree.TypeCheckAndRequire(s.builder.ctx, expr, s.builder.semaCtx, desired, s.context.String())
+	texpr, err := tree.TypeCheckAndRequire(s.builder.ctx, expr, s.builder.semaCtx, desired, s.context.String(), true, mutSuffix)
```

Comment:
I don't think we can allow assignment casts here in all cases. But, taking a step back, I'm curious why the optbuilder changes were necessary. What was failing before this change?

Reply:
After making the table with differing column and default types, trying to insert into the table would result in an error. This was caused by the condition in TypeCheckRequire
`if typ := typedExpr.ResolvedType(); !(typ.Equivalent(required) || typ.Family() == types.UnknownFamily) {`
I changed the function in opt builder because I noticed the mutation suffix could be passed here to be used in TypeCheckRequire. 
Would specifying that the mutation suffix has to be default or update narrow it down to only the cases we want?

Reply:
Ok, I think the change required in optbuilder was simpler, though tricky to figure out. I put up a PR: https://github.com/cockroachdb/cockroach/pull/81863. If you rebase your changes on that, you shouldn't need to make any changes in optbuilder.